### ISSUE TITLE
Linear decay scheduler

### ISSUE THREAD

**Issue by kylebgorman** [CUNY-CL/yoyodyne#28]
We only have one decaying scheduler right now: the goofy warmup+inverse square decay strategy from the original transformer paper. I propose that we also add a linear decay to `schedulers.py`.

@Adamits I think this is in your wheelhouse so I have assigned it to you; let me know if that's a problem.

**Comment by Adamits** [comment]
Sounds good! I have a few things for this project i'll have time to do after the ACL deadline tonight!

**Comment by Adamits** [comment]
I am looking at this now. I think we want to do this by adding pytorch schedulers to [this factory](https://github.com/CUNY-CL/yoyodyne/blob/3b1bf93c38dee004805912038c8ba6c40410f94b/yoyodyne/models/base.py#L300), and we only need to add to schedulers.py in the case that we want to wrap something in custom code.

Then, we can remove `warmup_steps` as an arg to the models, and just take a `schedulers_kwargs` dictionary. These will just be passed along to the schedulers, and we can even just defer to pytorch docs/exceptions for what should be passed through.

I am looking at the argparse docs now -- do they have a default way of grouping certain arguments in such a way that after parsing args, we get a specific dictionary of scheduler_kwargs? Does this implementation sound ok to you?

**Comment by kylebgorman** [comment]
> I am looking at this now. I think we want to do this by adding pytorch schedulers to [this factory](https://github.com/CUNY-CL/yoyodyne/blob/3b1bf93c38dee004805912038c8ba6c40410f94b/yoyodyne/models/base.py#L300), and we only need to add to schedulers.py in the case that we want to wrap something in custom code.

Following so far.

> Then, we can remove `warmup_steps` as an arg to the models, and just take a `schedulers_kwargs` dictionary. These will just be passed along to the schedulers, and we can just defer to pytorch docs/exceptions for what should be passed through, though I guess we need to add arguments that differ (e.g. for linear decay we need a start factor and end factor) for every scheduler (same would be true for more optimizers), as we add them.

That would be a major improvement.

> I am looking at the argparse docs now -- do they have a default way of grouping certain arguments in such a way that after parsing args, we get a specific dictionary of scheduler_kwargs? Does this implementation sound ok to you @kylebgorman ? Or else, I can just make that dict after we parse the args.

That sounds fine to me. You should probably create a function in `schedulers.py` with the signature `def add_argparse_args(parser: argparse.ArgumentParser) -> None: ...` and then call it in `train.py`, like we do for the other module-flag connections.

**Comment by kylebgorman** [comment]
I could imagine us enabling all the PyTorch schedulers as [there are a lot of them](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate) and some are obscure. But I think even just adding linear decay would go a long way.

**Comment by Adamits** [comment]
Right! I am just doing linear decay for this one (I imagine LSTM can benefit from the right scheduler), but it will be nice to have a framework for adding any others later.

**Comment by Adamits** [comment]
> That sounds fine to me. You should probably create a function in schedulers.py with the signature def add_argparse_args(parser: argparse.ArgumentParser) -> None: ... and then call it in train.py, like we do for the other module-flag connections.

Thanks! I will follow this.

**Comment by kylebgorman** [comment]
Closed in #31.
